File: cpp/src/price_level.cpp
#include "lumina_lob/price_level.hpp"

#include <algorithm>

namespace lumina_lob {

Order& PriceLevel::append(std::unique_ptr<Order> order) {
    total_qty_ += order->remaining_qty();
    order_count_ += 1;
    orders_.push_back(std::move(order));
    return *orders_.back();
}

bool PriceLevel::remove(const Order& order) {
    auto it = std::find_if(orders_.begin(), orders_.end(),
                           [&order](const std::unique_ptr<Order>& ptr) { return ptr.get() == &order; });
    if (it == orders_.end()) {
        return false;
    }
    total_qty_ -= (*it)->remaining_qty();
    order_count_ -= 1;
    orders_.erase(it);
    return true;
}

int64_t PriceLevel::fill(int64_t amount) {
    if (amount <= 0) {
        return 0;
    }
    int64_t remaining = amount;
    while (remaining > 0 && !orders_.empty()) {
        Order* front = orders_.front().get();
        int64_t can_fill = std::min(front->remaining_qty(), remaining);
        front->fill(can_fill);
        total_qty_ -= can_fill;
        remaining -= can_fill;
        if (front->is_filled()) {
            orders_.pop_front();
            order_count_ -= 1;
        }
    }
    return amount - remaining;
}

int64_t PriceLevel::reduce(Order& order, int64_t new_qty) {
    int64_t removed = order.reduce_qty(new_qty);
    total_qty_ -= removed;
    return removed;
}

bool PriceLevel::is_empty() const noexcept {
    return orders_.empty();
}

} // namespace lumina_lob

File: lumina_lob/core/price_level.py
"""Price level: queue of orders at same price using doubly-linked list."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from .order import Order


@dataclass
class PriceLevel:
    """Orders at one price level. FIFO queue via linked list."""

    price: int
    head: Optional[Order] = field(default=None, init=False)
    tail: Optional[Order] = field(default=None, init=False)
    total_qty: int = field(default=0, init=False)
    order_count: int = field(default=0, init=False)

    def append(self, order: Order) -> None:
        """Add order to tail."""
        if self.tail is None:
            self.head = self.tail = order
            order.prev = order.next = None
        else:
            order.prev = self.tail
            order.next = None
            self.tail.next = order
            self.tail = order
        self.total_qty += order.remaining_qty
        self.order_count += 1

    def remove(self, order: Order) -> bool:
        """Remove order from queue. Return True if found."""
        if order.prev is None and order.next is None and self.head is not order:
            return False
        if order.prev:
            order.prev.next = order.next
        else:
            self.head = order.next
        if order.next:
            order.next.prev = order.prev
        else:
            self.tail = order.prev
        self.total_qty -= order.remaining_qty
        self.order_count -= 1
        order.prev = order.next = None
        return True

    def fill(self, amount: int) -> int:
        """Fill from front of queue. Return filled amount."""
        remaining = amount
        while remaining > 0 and self.head:
            front = self.head
            can_fill = min(front.remaining_qty, remaining)
            front.fill(can_fill)
            self.total_qty -= can_fill
            remaining -= can_fill
            if front.is_filled:
                self.remove(front)
        return amount - remaining

    def reduce(self, order: Order, new_qty: int) -> int:
        """Reduce order qty at this level. Return removed amount."""
        removed = order.reduce_qty(new_qty)
        self.total_qty -= removed
        return removed

    def is_empty(self) -> bool:
        return self.head is None

    def __iter__(self):
        node = self.head
        while node:
            yield node
            node = node.next

    def __len__(self) -> int:
        return self.order_count
